app/src/main/java/com/github/epfl/meili/notifications/FirebaseNotificationService.kt

```diff
@@ -0,0 +1,124 @@
+package com.github.epfl.meili.notifications
+
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.app.NotificationManager.IMPORTANCE_HIGH
+import android.content.Context
+import android.content.SharedPreferences
+import android.graphics.Color
+import android.os.Build
+import android.util.Log
+import androidx.annotation.RequiresApi
+import androidx.core.app.NotificationCompat
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.ViewModelStoreOwner
+import com.github.epfl.meili.MainApplication
+import com.github.epfl.meili.R
+import com.github.epfl.meili.auth.Auth
+import com.github.epfl.meili.database.FirestoreDatabase
+import com.github.epfl.meili.models.Token
+import com.github.epfl.meili.models.User
+import com.github.epfl.meili.util.MeiliViewModel
+import com.google.firebase.database.DatabaseException
+import com.google.firebase.messaging.FirebaseMessaging
+import com.google.firebase.messaging.FirebaseMessagingService
+import com.google.firebase.messaging.RemoteMessage
+import kotlin.random.Random
+
+/**
+ * Custom service to receive notifications
+ */
+class FirebaseNotificationService() : FirebaseMessagingService() {
+
+    private lateinit var notificationManager: NotificationManager
+
+    companion object {
+        private const val CHANNEL_ID = "my_channel"
+
+        var sharedPref: SharedPreferences? = null
+
+        var token: String?
+            get() {
+                return sharedPref?.getString("token", "")
+            }
+            set(value) {
+                sharedPref?.edit()?.putString("token", value)?.apply()
+            }
+
+        fun registerToken(viewModelStoreOwner: ViewModelStoreOwner, user: User?) {
+            val tokenViewModel =
+                ViewModelProvider(viewModelStoreOwner).get(MeiliViewModel::class.java) as MeiliViewModel<Token>
+
+            tokenViewModel.initDatabase(FirestoreDatabase("token", Token::class.java))
+
+
+            FirebaseMessaging.getInstance().token.addOnSuccessListener {
+                token = it
+                if (user != null && it != null) {
+                    try {
+                        tokenViewModel.addElement(
+                            user!!.uid,
+                            Token(it)
+                        )
+                    } catch (e: Exception) {
+                        Log.e("MapActivity", "token already registered")
+                    }
+                }
+
+            }
+        }
+    }
+
+    override fun onNewToken(newToken: String) {
+        super.onNewToken(newToken)
+        token = newToken
+    }
+
+    override fun onMessageReceived(message: RemoteMessage) {
+        super.onMessageReceived(message)
+
+        Log.d("notif", "received message notification")
+        //Uncomment this code to make clicking notifications open activity
```

Comment:
Could you add an explanation of why we couldn't/chose to not do it.

Reply:
I added the things needed to make the commented code work. Maybe this way it's clearer why I couldn't do it 👍 .